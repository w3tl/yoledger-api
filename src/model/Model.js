import { ObjectId } from 'mongodb';

export default class Model {
  constructor(db, userId) {
    this.db = db;
    this.userId = userId;
  }

  async init() {
    const validator = {
      $jsonSchema: this.schema,
    };
    const collections = await this.db.collections();
    if (collections.indexOf(this.collName) === -1) {
      await this.db.createCollection(this.collName, {
        validator,
      });
    } else {
      await this.db.command({
        collMod: this.collName,
        validator,
      });
    }
  }

  get collection() { return this.db.collection(this.collName); }

  clear() {
    return this.collection.deleteMany({ userId: this.userId });
  }

  deleteOne(filter = {}, options) {
    filter.userId = this.userId; // eslint-disable-line no-param-reassign
    return this.collection.deleteOne(filter, options);
  }

  find(query = {}, options) {
    query.userId = this.userId; // eslint-disable-line no-param-reassign
    return this.collection.find(query, options);
  }

  findOne(query = {}, options) {
    query.userId = this.userId; // eslint-disable-line no-param-reassign
    return this.collection.findOne(query, options);
  }

  findById(id, options) {
    return this.collection.findOne({ _id: ObjectId(id) }, options);
  }

  findOneAndDelete(filter = {}, options) {
    filter.userId = this.userId; // eslint-disable-line no-param-reassign
    return this.collection.findOneAndDelete(filter, options);
  }

  insertOne(doc = {}, options) {
    doc.userId = this.userId; // eslint-disable-line no-param-reassign
    return this.collection.insertOne(doc, options);
  }

  insertMany(docs, options) {
    const docsWithUserId = docs.map(doc => ({ ...doc, ...{ userId: this.userId } }));
    return this.collection.insertMany(docsWithUserId, options);
  }

  updateOne(filter = {}, update, options) {
    filter.userId = this.userId; // eslint-disable-line no-param-reassign
    return this.collection.updateOne(filter, update, options);
  }

  aggregate(pipeline, options) {
    let [firstPipe, ...pipes] = pipeline; // eslint-disable-line prefer-const
    if ('$match' in firstPipe) {
      firstPipe.$match.userId = this.userId;
      return this.collection.aggregate([firstPipe, ...pipes], options);
    }
    firstPipe = { $match: { userId: this.userId } };
    return this.collection.aggregate([firstPipe, ...pipeline], options);
  }
}
