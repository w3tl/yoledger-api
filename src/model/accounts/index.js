import Model from '../Model';

export const accountTypes = ['ASSET', 'EXPENSE', 'INCOME', 'VENDOR'];

const schema = {
  type: 'object',
  required: ['name', 'type', 'userId'],
  additionalProperties: false,
  properties: {
    _id: {
      bsonType: 'objectId',
    },
    name: {
      type: 'string',
    },
    balance: {
      type: 'number',
    },
    type: {
      enum: accountTypes,
    },
    userId: {
      type: 'string',
    },
  },
};

export default class Account extends Model {
  constructor(db, userId) {
    super(db, userId);
    this.schema = schema;
    this.collName = 'accounts';
  }

  create({
    name, type, balance = 0,
  }) {
    return super.insertOne({
      name,
      type,
      balance,
    });
  }

  addAmount(name, amount, session) {
    return super.updateOne(
      { name, userId: this.userId },
      { $inc: { balance: amount } },
      { session },
    );
  }

  findByName(name, ...props) {
    return super.findOne({ name }, ...props);
  }
}
