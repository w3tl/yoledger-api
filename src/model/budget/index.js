import Model from '../Model';
import Account from '../accounts';
import schema from './schema';

export default class Budget extends Model {
  constructor(db, userId) {
    super(db, userId);
    this.collName = 'budget';
    this.schema = schema;
  }

  async upsertBudget({ account, date, amount }) {
    const accountModel = new Account(this.db, this.userId);
    const accountDoc = await accountModel.findByName(account, {
      projection: { _id: 0, name: 1, type: 1 },
    });
    if (!accountDoc) {
      throw new Error(`Account "${account}" not found`);
    }
    const periodDate = typeof date === 'string'
      ? new Date(date) : date;
    const { upsertedId, modifiedCount } = await super.updateOne({
      date: periodDate, account: accountDoc,
    }, {
      $set: { amount },
    }, {
      upsert: true,
    });
    return upsertedId || modifiedCount;
  }
}
