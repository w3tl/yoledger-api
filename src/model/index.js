import Account from './accounts';
import Transaction from './transactions';
import Budget from './budget';
import User from './user';

export {
  User,
  Account,
  Transaction,
  Budget,
};
