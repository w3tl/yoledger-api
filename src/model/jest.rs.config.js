module.exports = {
  testEnvironment: 'node',
  notify: true,
  transform: {
    '^.+\\.js?$': 'babel-jest',
  },
  globalSetup: './setup.js',
  globalTeardown: './teardown.js',
};
