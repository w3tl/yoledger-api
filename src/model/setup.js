const { ReplSet } = require('mongodb-topology-manager');

module.exports = async function setup() {
  const replSet = new ReplSet('mongod', [{
    options: {
      bind_ip: 'localhost', port: 31000, dbpath: `${__dirname}/data/db/31000`,
    },
  }], {
    replSet: 'rs',
  });

  await replSet.purge();
  await replSet.start();

  global.__MONGOD__ = replSet;
};
