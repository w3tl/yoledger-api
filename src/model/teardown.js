module.exports = async function teardown() {
  await global.__MONGOD__.stop();
};
