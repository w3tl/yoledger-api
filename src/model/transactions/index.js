import Model from '../Model';
import Account from '../accounts';
import schema from './schema';

export default class Transaction extends Model {
  constructor(db, userId) {
    super(db, userId);
    this.collName = 'transactions';
    this.schema = schema;
  }

  async post({
    source, destination, amount, createdAt = new Date(),
  }, session) {
    const Accounts = new Account(this.db, this.userId);
    const sourceAccount = await Accounts.findByName(source);
    const destAccount = await Accounts.findByName(destination);

    if (!sourceAccount) {
      throw new Error(`Account "${source}" not found`);
    }
    if (!destAccount) {
      throw new Error(`Account "${destination}" not found`);
    }

    const transObj = {
      source: { name: sourceAccount.name, type: sourceAccount.type },
      destination: { name: destAccount.name, type: destAccount.type },
      amount,
      createdAt,
    };
    // TODO: apply to budget
    const insertedId = await this.postFunc(transObj, session);
    return insertedId;
  }

  async postFunc(transObj, session) {
    try {
      session.startTransaction();
      const { insertedId } = await super.insertOne(transObj, { session });
      const Accounts = new Account(this.db, this.userId);
      await Accounts.addAmount(transObj.source.name, -transObj.amount, session);
      await Accounts.addAmount(transObj.destination.name, transObj.amount, session);
      await session.commitTransaction();
      return insertedId;
    } catch (error) {
      await session.abortTransaction();
      if (error.errorLabels && error.errorLabels.indexOf('TransientTransactionError') >= 0) {
        // console.log('TransientTransactionError, retrying transaction ...');
        return this.postFunc(transObj, session);
      }
      throw error;
    }
  }

  async unpost(_id, session) {
    try {
      session.startTransaction();
      const { value } = await super.findOneAndDelete(
        { _id },
        { projection: { source: 1, destination: 1, amount: 1 } },
      );
      if (!value) { throw new Error('Transaction not found'); }
      const { source, destination, amount } = value;
      const Accounts = new Account(this.db, this.userId);
      await Accounts.addAmount(source.name, amount, session);
      await Accounts.addAmount(destination.name, -amount, session);
      // TODO: apply to budget
    } catch (e) {
      await session.abortTransaction();
      throw e;
    }

    await session.commitTransaction();
  }

  async update(_id, {
    source, destination, amount, createdAt,
  }, session) {
    const trans = await super.findOne({ _id });
    if (!trans) { throw new Error('Transaction not found'); }

    if (source || destination || amount) {
      await this.unpost(_id, session);
      const transObj = {
        source: source || trans.source.name,
        destination: destination || trans.destination.name,
        amount: amount || trans.amount,
        createdAt: createdAt || trans.createdAt,
      };

      const insertedId = await this.post(transObj, session);
      return insertedId;
    }
    if (createdAt) {
      await super.updateOne({ _id }, { $set: { createdAt } });
      return _id;
    }
    return null;
  }
}
