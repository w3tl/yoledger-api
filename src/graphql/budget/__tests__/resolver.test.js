jest.mock('mongodb');

import { dataloaders } from '../../index';
import {
  Query, Mutation, Budgets,
} from '../resolver';
import { Account, Budget as BudgetModel } from '../../../model';

describe('budget resolver', () => {
  const { connection } = require('mongodb');
  const context = {
    dataloaders: dataloaders(connection),
    models: {
      budgetModel: new BudgetModel(connection.db(), 'admin'),
      accountModel: new Account(connection.db(), 'admin'),
    },
    userId: 'admin',
  };

  describe('Query', () => {
    test('bugdets should return accounts', async () => {
      const result = await Query.budgets(null, {
        dateStart: new Date('2018-05-10'),
        dateEnd: new Date('2018-06-25'),
      }, context);
      expect(result).toMatchSnapshot();
    });

    test('budget.account should return an account', async () => {
      const result = await Query.budget(null, { account: { name: 'Food' } }, context);
      expect(result).toMatchSnapshot();
    });
  });

  describe('Mutation', () => {
    test('upsertBudget', async () => {
      const result = await Mutation.upsertBudget(null, {
        input: {
          account: 'Food',
          date: new Date('2018-05-10'),
          amount: 564,
        },
      }, context);
      expect(result).toMatchSnapshot();
    });
  });
});
