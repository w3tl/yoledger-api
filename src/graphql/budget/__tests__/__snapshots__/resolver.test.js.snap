// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`budget resolver Mutation upsertBudget 1`] = `
Object {
  "budget": Object {
    "_id": "2",
    "account": Object {
      "name": "Food",
      "type": "EXPENSE",
    },
    "amount": 55,
    "date": 2018-05-10T00:00:00.000Z,
  },
  "success": "2",
}
`;

exports[`budget resolver Query budget.account should return an account 1`] = `Array []`;

exports[`budget resolver Query bugdets should return accounts 1`] = `
Object {
  "accounts": Array [
    Object {
      "_id": "3",
      "name": "Food",
      "type": "EXPENSE",
    },
    Object {
      "_id": "4",
      "name": "Train",
      "type": "EXPENSE",
    },
  ],
}
`;
